packages/node/src/indexer/fetch.service.ts

```diff
@@ -100,41 +186,106 @@ export class FetchService implements OnApplicationShutdown {
   latestProcessed(height: number): void {
     this.latestProcessedHeight = height;
   }
-
   async startLoop(initBlockHeight: number): Promise<void> {
+    await Promise.all([
+      this.fillNextBlockBuffer(initBlockHeight),
+      this.fillBlockBuffer(),
+    ]);
+  }
+
+  async fillNextBlockBuffer(initBlockHeight: number): Promise<void> {
     if (isUndefined(this.latestProcessedHeight)) {
       this.latestProcessedHeight = initBlockHeight - 1;
     }
     await this.fetchMeta(initBlockHeight);
     // eslint-disable-next-line no-constant-condition
+    let startBlockHeight = initBlockHeight;
+
     while (!this.isShutdown) {
-      const [startBlockHeight, endBlockHeight] =
-        this.nextBlockRange(initBlockHeight) ?? [];
-      if (isUndefined(startBlockHeight)) {
+      if (this.nextBlockBuffer.size > this.bufferAllowSize) {
```

Comment:
There are other condition branches you need to take into consideration.
like latestProcessedHeight == latest finalised block height